crates/sui-json-rpc/src/metrics.rs

```diff
@@ -0,0 +1,169 @@
+    IntCounterVec,
+};
+use serde::Deserialize;
+use tokio::runtime::Handle;
+use tokio::sync::Mutex;
+use tokio::time::Instant;
+use tower::Layer;
+
+#[derive(Debug, Clone)]
+pub struct MetricsLayer {
+    metrics: Arc<Metrics>,
+}
+impl MetricsLayer {
+    pub fn new(registry: &prometheus::Registry) -> Self {
+        let metrics = Arc::new(Metrics {
+            requests_by_route: register_int_counter_vec_with_registry!(
+                "rpc_requests_by_route",
+                "Number of requests by route",
+                &["route"],
+                registry,
+            )
+            .unwrap(),
+            req_latency_by_route: register_histogram_vec_with_registry!(
+                "req_latency_by_route",
+                "Latency of a request by route",
+                &["route"],
+                LATENCY_SEC_BUCKETS.to_vec(),
+                registry,
+            )
+            .unwrap(),
+            errors_by_route: register_int_counter_vec_with_registry!(
+                "errors_by_route",
+                "Number of errors by route",
+                &["route"],
+                registry,
+            )
+            .unwrap(),
+        });
+
+        Self { metrics }
+    }
+}
+
+impl<S> Layer<S> for MetricsLayer {
+    type Service = JsonRpcMetricService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        JsonRpcMetricService::new(inner, self.metrics.clone())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct JsonRpcMetricService<S> {
+    inner: Arc<Mutex<S>>,
+    metrics: Arc<Metrics>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Metrics {
+    /// Counter of requests, route is a label (ie separate timeseries per route)
+    requests_by_route: IntCounterVec,
+    /// Request latency, route is a label
+    req_latency_by_route: HistogramVec,
+    /// Failed requests by route
+    errors_by_route: IntCounterVec,
+}
+
+const LATENCY_SEC_BUCKETS: &[f64] = &[
+    0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5, 1., 2.5, 5., 10., 20., 30., 60., 90.,
+];
+
+impl<S> JsonRpcMetricService<S> {
+    pub fn new(inner: S, metrics: Arc<Metrics>) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(inner)),
+            metrics,
+        }
+    }
+}
+
+impl<S> Service<Request<Body>> for JsonRpcMetricService<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>> + Send + Sync + 'static,
+    S::Response: 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let handle = Handle::current();
+        let _ = handle.enter();
+        let mut v = futures::executor::block_on(self.inner.lock());
```

Comment:
You cannot use another executor inside of a tokio runtime. This can lead to deadlocking. The way that tower services are supposed to work is that you have an owned service, not a wrapper around a shared one with Mutexes

Reply:
Generally services end up being clone-able and they're just cloned to handle concurrent requests.

Reply:
Sadly jsonrpsee's TowerService is not clone-able :( , I will fix this upstream

Reply:
PR created upstream, https://github.com/paritytech/jsonrpsee/pull/950

I am using a forked version instead until next jsonrpsee release